modules/lsp.nix

```diff
@@ -0,0 +1,159 @@
+{
+  lib,
+  config,
+  ...
+}:
+let
+  inherit (lib) mkOption types;
+  inherit (lib.nixvim) toLuaObject;
+
+  cfg = config.lsp;
+
+  serverType = types.submodule {
+    options = {
+      enable = lib.mkEnableOption "the language server";
+
+      activate = lib.mkOption {
+        type = types.bool;
+        description = ''
+          Whether to call `vim.lsp.enable()` for this server.
+        '';
+        default = true;
+        example = false;
+      };
+
+      package = lib.mkOption {
+        type = with types; nullOr package;
+        default = null;
+        description = ''
+          Package to use for this language server.
+
+          Alternatively, the language server should be available on your `$PATH`.
+        '';
+      };
+
+      config = mkOption {
+        type = with types; attrsOf anything;
+        description = ''
+          Configurations for each language server.
+        '';
+        default = { };
+        example = {
+          cmd = [
+            "clangd"
+            "--background-index"
+          ];
+          root_markers = [
+            "compile_commands.json"
+            "compile_flags.txt"
+          ];
+          filetypes = [
+            "c"
+            "cpp"
+          ];
+        };
+      };
+    };
+  };
+in
+{
+  options.lsp = {
+    luaConfig = lib.mkOption {
+      type = lib.types.pluginLuaConfig;
+      default = { };
+      description = ''
+        Lua code configuring LSP.
+      '';
+    };
+
+    inlay_hint = {
+      enable = lib.mkEnableOption "inlay hints globally";
+    };
+
+    servers = mkOption {
+      type = types.attrsOf serverType;
+      description = ''
+        LSP servers to enable and/or configure.
+
+        This option is implemented using neovim's `vim.lsp` lua API.
+        If you prefer to use [nvim-lspconfig], see [`plugins.lsp`].
```

Comment:
At some point, we should tweak this wording to suggest users use _both_ `lsp` and `plugins.lsp` together.